File: solutions/java/resistor-color-trio/4/src/main/java/ResistorColorTrio.java
class ResistorColorTrio {

    int colorCode(String color) {
        String colorLower = color.toLowerCase();
        switch (colorLower) {
            case "black":
                return 0;
            case "brown":
                return 1;
            case "red":
                return 2;
            case "orange":
                return 3;
            case "yellow":
                return 4;
            case "green":
                return 5;
            case "blue":
                return 6;
            case "violet":
                return 7;
            case "grey":
                return 8;
            case "white":
                return 9;
            default:
                throw new IllegalArgumentException("Unknown color: " + color);
        }
    }
    
    int value(String[] colors) {
        if (colors.length < 2) {
            throw new IllegalArgumentException("Invalid number of arguments.");
        }
        return colorCode(colors[0]) * 10 + colorCode(colors[1]);
    }

    String label(String[] colors) {
        if (colors.length < 3) {
            throw new IllegalArgumentException("Invalid number of arguments.");
        }
        int value = value(colors);
        int zeros = colorCode(colors[2]);
        double resistence = value * Math.pow(10, zeros);

        if (resistence >= 1e9) {
            resistence /= 1e9;
            String formatdecimal = resistence % 1 == 0 ? "%.0f" : "%.1f"; 
            return String.format(formatdecimal + " gigaohms", resistence);
        }
        if (resistence >= 1e6) {
            resistence /= 1e6;
            String formatdecimal = resistence % 1 == 0 ? "%.0f" : "%.1f"; 
            return String.format(formatdecimal + " megaohms", resistence);
        }
        else if ( resistence >= 1e3) {
            resistence /= 1e3;
            String formatdecimal = resistence % 1 == 0 ? "%.0f" : "%.1f"; 
            return String.format(formatdecimal + " kiloohms", resistence);
        }
        return String.format("%.0f ohms", resistence);
    }

}

File: solutions/python/resistor-color-trio/2/resistor_color_trio.py
from math import log10, floor

COLOR_VALUES = [
    'black', 'brown', 'red', 'orange', 'yellow', 'green', 'blue', 'violet', 'grey', 'white'
]

def color_code(color):
    return COLOR_VALUES.index(color)

def colors():
    return COLOR_VALUES

def value(colors):
    codes = [color_code(item) for item in colors]
    suma = 0
    for i, item in enumerate(reversed(codes), start=0):
        suma += item * 10 ** i
    return suma

PREFIX_LIST = ["", "kilo", "mega", "giga"]

def ohm_value(value):
    if value == 0:
        return "0 ohms"
    log10value = log10(value)
    log10div3 = floor(log10value) // 3
    new_value = value // (10 ** (3 * log10div3))
    return f"{new_value} {PREFIX_LIST[log10div3]}ohms"

def label(colors):
    *first_colors, last_color = colors[:3]
    value_colors = value(first_colors)
    last_code = color_code(last_color)
    resistence_value = value_colors * 10 ** last_code
    return ohm_value(resistence_value)
